Src/library/net/Clib/network.c

```diff
@@ -775,11 +779,162 @@ EIF_INTEGER c_put_stream_noexception(EIF_INTEGER fd, EIF_POINTER stream_pointer,
+	 */
+{
+#define EIFNET_BUFFSIZE 4096
+	int fd;
+	int retval_read, retval_write;
+	int nb_bytes_sent, l_read_remain;
+	int n, l_write_remain;
+	char buf[EIFNET_BUFFSIZE];
+
+#ifdef EIF_WINDOWS
+	fd = _fileno(f);
+#else
+	fd = fileno(f);
+#endif
+	nb_bytes_sent = 0;
+	l_read_remain = length;
+	do {
+			/* Read chunk of bytes */
+		if (l_read_remain < EIFNET_BUFFSIZE) {
+			n = l_read_remain;
+		} else {
+			n = EIFNET_BUFFSIZE;
+		}
+#ifdef EIF_WINDOWS		
+		retval_read = (int)_read(fd, buf, n);
+#else
+		retval_read = (int) read(fd, buf, n);
+#endif
+		if (retval_read < 0) {
+				/* error while reading file */
+			return (EIF_INTEGER) retval_read;
+		} else {
+			l_read_remain = l_read_remain - retval_read;
+		}
+			/* Send read bytes */
+		l_write_remain = retval_read;
+		do {
+			retval_write = (int) c_put_stream_noexception(out_fd, buf, l_write_remain);
+			if (retval_write < 0) {
+					/* error while writing into socket */
+				return (EIF_INTEGER) retval_write;
+			} else {
+				l_write_remain = l_write_remain - retval_write;
+				nb_bytes_sent = nb_bytes_sent + retval_write;
+			}
+		} while (l_write_remain > 0);
+
+	} while (retval_read || retval_write);
+
+		/* If this code is reached, no file or network error occurred
+		 * thus return the number of bytes read and sent.
+		 */
+	return (EIF_INTEGER) nb_bytes_sent;
+#undef EIFNET_BUFFSIZE
+}
+
+EIF_INTEGER c_sendfile_noexception(EIF_INTEGER out_fd, FILE* f, EIF_INTEGER length)
+	/* transmission of file content from file `f` of size length through socket out_fd.
+	 * if `length` is 0, then send the whole file.
+	 *	NO exception is raised, and eventual error is return as result!
+	 */
+{
+#ifdef EIF_WINDOWS
+#	ifndef LPTRANSMIT_FILE_BUFFERS 
+#		define LPTRANSMIT_FILE_BUFFERS  void* 
+#	endif
+#	ifndef TF_WRITE_BEHIND 
+#		define TF_WRITE_BEHIND     0x04
+#	endif
+#	ifndef TF_REUSE_SOCKET 
+#		define TF_REUSE_SOCKET     0x20
+#	endif
+	
+	static BOOL (PASCAL * _TransmitFile)(SOCKET, HANDLE, DWORD, DWORD, LPOVERLAPPED, LPTRANSMIT_FILE_BUFFERS, DWORD) = NULL;
+	
+	if (_TransmitFile) {
+		BOOL retval;
+		BOOL l_continue;
+		HANDLE hFile = (HANDLE)_get_osfhandle(fileno(f));
+		OVERLAPPED ovlp;
+		DWORD len;
+		len = (DWORD) length;
+		memset(&ovlp, 0, sizeof(OVERLAPPED));
+		do {
+			retval = _TransmitFile((EIF_SOCKET_TYPE) out_fd, (HANDLE) hFile, len, 
+					(DWORD) 0 /* Use default Windows block size */, (LPOVERLAPPED) &ovlp, (LPTRANSMIT_FILE_BUFFERS) NULL,(DWORD) TF_REUSE_SOCKET | TF_WRITE_BEHIND);
+			l_continue = FALSE;
+			if (retval < 0) {
+				int errcode;
+				errcode = WSAGetLastError();
+				if ((errcode == WSA_IO_PENDING) || (errcode == ERROR_IO_PENDING)) {
+					l_continue = TRUE;
+				} else {
+					l_continue = FALSE;
+				}
+			}
+		} while (l_continue);
+		if (retval) {
+				/* Fix me, find a way to get the number of bytes sent */
+			return (EIF_INTEGER) ovlp.InternalHigh;
+		} else {
+			return (EIF_INTEGER) -1;
+		}
+	} else {
+		static BOOL _TransmitFile_searched = FALSE;
```

Comment:
Was reading the doc and you need to create an EVent in the OVERLAPPED structure and if you havE WSA_IO_PENDING, you just wait until the event is signaled. That way you are sure it is synchronous.